### PR TITLE
Implemented BlackBerry10 push support

### PR DESCRIPTION
**Pull request by jamesjhedley**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -320,7 +329,7 @@ if ( device.platform == 'android' || device.platform == 'Android' || device.plat
 }
 ```
 
-On success, you will get a call to tokenHandler (iOS), onNotification (Android and Amazon Fire OS), or onNotificationWP8 (WP8), allowing you to obtain the device token or registration ID, or push channel name and Uri respectively. Those values will typically get posted to your intermediary push server so it knows who it can send notifications to.
+On success, you will get a call to tokenHandler (iOS), onNotification (Android and Amazon Fire OS), onNotificationWP8 (WP8) or success (Blackberry10), allowing you to obtain the device token or registration ID, or push channel name and Uri respectively. Those values will typically get posted to your intermediary push server so it knows who it can send notifications to.
```
Comment (bryanhiggins) [inline README.md]: shouldn't 'success' be 'successHandler' ?

File: README.md [hunk 2]
```diff
@@ -328,6 +337,9 @@ On success, you will get a call to tokenHandler (iOS), onNotification (Android a
 
 - **Android**: If you have not already done so, you'll need to set up a Google API project, to generate your senderID. [Follow these steps](http://developer.android.com/guide/google/gcm/gs.html) to do so. This is described more fully in the **Testing** section below. In this example, be sure and substitute your own senderID. Get your senderID by signing into to your [google dashboard](https://code.google.com/apis/console/). The senderID is found at *Overview->Dashboard->Project Number*.
 
+- **BlackBerry10**: "ecb" MUST be provided for this platform in order to get push notifications. The invokeTargetId parameter above is optional. By default an invoke target will be automatically added to the config and used when registering to invoke the app. You can optionally use this parameter to specify a different id for an invoke target maintained by you. Also note, if doing a public consumer (BIS) push, you need to manually add the _sys_use_consumer_push permission to config.xml. `<rim:permit system="true">_sys_use_consumer_push</rim:permit>`
```
Comment (bryanhiggins) [inline README.md]: This is confusing. I suggest linking to the WebWorks docs.

ecb - if this is omitted does no notification appear in the tray? or do I just not get the oninvoked payload?

'to the config' - config.xml?

'a different id for an invoke target maintained by you' - unclear what this means

File: README.md [hunk 3]
```diff
@@ -307,6 +306,16 @@ if ( device.platform == 'android' || device.platform == 'Android' || device.plat
...
     });
+} else if ( device.platform == 'blackberry10'){
+    pushNotification.register(
+    successHandler,
+    errorHandler,
+    {
+        invokeTargetId : "replace_with_invoke_target_id",//optional
+        appId: "replace_with_app_id", //remove for BES pushes
+        ppgUrl:"replace_with_ppg_url", //remove for BES pushes
+        ecb: "replace_with_push_notification_handler"
```
Comment (bryanhiggins) [inline README.md]: Why not name this "pushNotificationHandler" to match the function below?

File: plugin.xml [hunk 4]
```diff
@@ -125,8 +126,29 @@
...
+        <dependency id="com.blackberry.push" />
+        <dependency id="com.blackberry.invoked" />
+        <config-file target="www/config.xml" parent="/widget">
+            <feature name="PushPlugin" value="PushPlugin" />
+        </config-file>
+        <js-module src="www/blackberry10/PushPluginProxy.js" name="PushPluginProxy" >
+            <runs />
+        </js-module>
+        <config-file target="www/config.xml" parent="/widget">
+            <rim:invoke-target id="PushPluginInvokeId">
```
Comment (bryanhiggins) [inline plugin.xml]: Doesn't this need to be unique for each app?

For the BB plugin, we ask users to add this to config.xml manually (wrapped in a platform tag)

File: www/blackberry10/PushPluginProxy.js [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    register: function(success, error, args) {
+        var ops = args[0],
+            ecb = ops.ecb,
+            simChangeCallback = ops.simChangeCallback,
+            pushTransportReadyCallback = ops.pushTransportReadyCallback,
+            launchApplicationOnPush = ops.launchApplicationOnPush,
+            pushPayLoad,
+            pushCallback;
+
+        ops.invokeTargetId = ops.invokeTargetId || "PushPluginInvokeId";
```
Comment (bryanhiggins) [inline www/blackberry10/PushPluginProxy.js]: We should not provide a default value

File: www/blackberry10/PushPluginProxy.js [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+module.exports = {
+
+    register: function(success, error, args) {
+        var ops = args[0],
+            ecb = ops.ecb,
+            simChangeCallback = ops.simChangeCallback,
+            pushTransportReadyCallback = ops.pushTransportReadyCallback,
+            launchApplicationOnPush = ops.launchApplicationOnPush,
+            pushPayLoad,
+            pushCallback;
```
Comment (bryanhiggins) [inline www/blackberry10/PushPluginProxy.js]: pushCallback isn't used outside of invoked handler (it should be scoped there)

File: www/blackberry10/PushPluginProxy.js [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+                      + "programming error.");
+            } else {
+                error("Error: Received error code (" + result + ") after "
+                      + "calling blackberry.push.PushService.create.");
+            }
+        }, simChangeCallback, pushTransportReadyCallback);
+    },
+
+    unregister: function(success, error, args) {
+        if (pushServiceObj) {
```
Comment (bryanhiggins) [inline www/blackberry10/PushPluginProxy.js:172]: We should remove the oninvoked listener here

**Comment by jamesjhedley** [comment]
All great suggestions. Thanks.
I have made the recommended changes.

**Comment by bryanhiggins** [comment]
Looks good!

**Comment by jamesjhedley** [comment]
Squashed.
Retested. BIS and BES working well.
I can go over it with you tomorrow as well if you want.

**Pull request merged**